pkg/fleet/installer/service/file.go

```diff
@@ -0,0 +1,161 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build !windows
+
+// Package service provides a way to interact with os services
+package service
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"syscall"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// fileMutator is a struct used to transform a file
+// creating backups, replacing original files and setting permissions
+// default permissions are root:root 0644
+type fileMutator struct {
+	path             string
+	pathTmp          string
+	pathBackup       string
+	transformContent func(existing []byte) ([]byte, error)
+	validateTemp     func() error
+	validateFinal    func() error
+}
+
+// newFileMutator creates a new fileMutator
+func newFileMutator(path string, transform func(existing []byte) ([]byte, error), validateTemp, validateFinal func() error) *fileMutator {
+	return &fileMutator{
+		path:             path,
+		pathTmp:          path + ".datadog.prep",
+		pathBackup:       path + ".datadog.backup",
+		transformContent: transform,
+		validateTemp:     validateTemp,
+		validateFinal:    validateFinal,
+	}
+}
+
+func (ft *fileMutator) mutate() (rollback func() error, err error) {
+	defer os.Remove(ft.pathTmp)
+
+	originalFileExists := true
+	// create backup and temporary file if the original file exists
+	if _, err := os.Stat(ft.path); err != nil {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ Question:** ‏Should we check for dir existence too?

Reply:
if it's possible that a file exist and not it's directory, we will fail anyways on the creation of the copies, which is ok too